File: Assignments/src/a7/p4D.cpp
#include <bits/stdc++.h>
using namespace std;

struct Envelope { int w, h, n, max, prev; };

bool env_sort (Envelope const& a, Envelope const& b) { return a.w < b.w; }
bool fits_in (Envelope a, Envelope b) { return a.w < b.w && a.h < b.h; }

int main() {
// READ IN
    int N, W, H;
    scanf("%d %d %d", &N, &W, &H);
    Envelope envs[N];
    for (int i = 0; i < N; i++) {
        scanf("%d %d", &envs[i].w, &envs[i].h);
        envs[i].n = i+1;
        envs[i].max = envs[i].w > W && envs[i].h > H ? 1 : 0;
    }
// DO LOGIC
    sort(envs, envs+N, &env_sort);
    int maxIndx = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < i; j++) {
            if (fits_in(envs[j], envs[i]) && envs[i].max <= envs[j].max && envs[j].max > 0) {
                envs[i].max = envs[j].max+1;
                envs[i].prev = j;
            }
        }
        if (envs[i].max > envs[maxIndx].max)
            maxIndx = i;
    }
    int maxVal = envs[maxIndx].max;
    printf("%d\n", maxVal);
    // get path
    if (maxVal > 0) {
        int current = maxIndx;
        int path[maxVal];
        for (int i = maxVal-1; i >= 0; i--) {
            path[i] = envs[current].n;
            current = envs[current].prev;
        }
        for (int i = 0; i < maxVal; i++)
            printf("%d ", path[i]);
        printf("\n");
    }
}

File: Assignments/src/a7/p4D.java
package a7;

import java.util.Arrays;
import java.util.Scanner;

public class p4D {
	private static class Envelope implements Comparable<Envelope> {
		int w, h, max, prev, n;
		Envelope(int width, int height, int number) {
			w = width;
			h = height;
			n = number;
		}
		public boolean fitsIn(Envelope a) { return a.w > w && a.h > h; }
		@Override
		public int compareTo(Envelope arg0) { return w-arg0.w; }
		@Override
		public String toString() { return n+": <"+max+"> ("+w+", "+h+")"; }
	}
	
	public static void main(String[] args) {
	// READ IN
	    int N, W, H;
	    Scanner in = new Scanner(System.in);
	    N = in.nextInt();
	    W = in.nextInt();
	    H = in.nextInt();
	    Envelope[] envelopes = new Envelope[N];
	    for (int i = 0; i < N; i++) { // read in envelopes and set max to 1 if they can fit the card
	    	envelopes[i] = new Envelope(in.nextInt(), in.nextInt(), i+1);
	    	envelopes[i].max = envelopes[i].w > W && envelopes[i].h > H ? 1 : 0;
	    }
	    in.close();
	// DO LOGIC
	    Arrays.sort(envelopes); // sort according to width
	    int maxIndx = 0;
	    for (int i = 0; i < N; i++) {
	    	for (int j = 0; j < i; j++) {
	    		if (envelopes[j].fitsIn(envelopes[i]) && envelopes[j].max >= envelopes[i].max && envelopes[j].max > 0) {
	    			envelopes[i].max = envelopes[j].max + 1;
	    			envelopes[i].prev = j;
	    		}
	    	}
	    	if (envelopes[i].max > envelopes[maxIndx].max)
	    		maxIndx = i;
	    }
	    // Get path (must read it backwards into an array from the prevs)
	    int current = maxIndx, maxVal = envelopes[maxIndx].max;
	    int[] path = new int[maxVal];
	    for (int i = maxVal-1; i >= 0; i--) {
	    	path[i] = envelopes[current].n;
	    	current = envelopes[current].prev;
	    }
	    // Print results
    	System.out.println(maxVal);
	    if (maxVal > 0)
		    for (int i = 0; i < path.length; i++)
		    	System.out.print(path[i]+" ");
	}

}
